neurom/core/features.py

```diff
@@ -0,0 +1,136 @@
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+'''Feature registrationg and retrieval'''
+
+import numpy as np
+
+from functools import partial
+from ..exceptions import NeuroMError
+from .types import tree_type_checker
+from . import iter_neurites
+from . import NeuriteType
+
+
+FEATURES = {'NEURITEFEATURES': {},
+            'NEURONFEATURES': {}, }
+
+
+def register_neurite_feature(name, func):
+    '''Register a feature to be applied to neurites
+
+    Parameters:
+        name: name of the feature, used for access via get() function.
+        func: single parameter function of a neurite.
+    '''
+    def _fun(neurites, neurite_type=NeuriteType.all):
+        '''Wrap neurite function from outer scope and map into list'''
+        return list(func(n)
+                    for n in iter_neurites(neurites, filt=tree_type_checker(neurite_type)))
+
+    register_feature('NEURITEFEATURES', name, _fun)
+
+
+def register_feature(namespace, name, func):
+    '''Register a feature to be applied
+
+    Parameters:
+        namespace(string):
+        name(string): name of the feature, used for access via get() function.
+        func(callable): single parameter function of a neurite.
+    '''
+    if name in FEATURES[namespace]:
+        raise NeuroMError('Attempt to hide registered feature %s' % name)
+
+    FEATURES[namespace][name] = func
+
+
+def feature(func=None, namespace=None, names=''):
+    '''feature decorator to automatically register the feature in the appropriate namespace'''
+
+    if func is None:
+        return partial(feature, namespace=namespace, names=names)
+
+    names = set(names.split()) | set([func.__name__, ])
+    for name in names:
+        register_feature(namespace, name, func)
+
+    return func
+
+
+def get(feature_name, obj, resolution_order=('NEURITEFEATURES', 'NEURONFEATURES', ), **kwargs):
+    '''Obtain a feature from a set of morphology objects
+
+    Parameters:
+        feature_name(string): feature to extract
+        obj: a neuron, population or neurite tree
+        **kwargs: parameters to forward to underlying worker functions
+
+    Returns:
+        features as a 1D or 2D numpy array.
+    '''
+
+    for name in resolution_order:
+        if feature_name in FEATURES[name]:
+            feat = FEATURES[name][feature_name]
+            break
+    else:
+        raise NeuroMError('Unable to find feature: %s' % feature_name)
```

Comment:
My concern is that this will be wrong for features that apply to both resolution orders, such as the ```principal_direction_extents``` that I suggested in #567.

Unless the user passes ```NEURONFEATURES``` as the ```resolution_order``` or swap the elements in the tuple, the feature will be assumed as a neurite feature and treat the cell as a collection of separate neurites, which gets the specific feature for a neuron object.

My rationale when I created the two feature categories was to have the ability to apply some features in both resolution orders, if it is required.